src/components/menu/menu-item/menu-item.component.js

```diff
@@ -1,33 +1,211 @@
+    if (!Events.composedPath(e).includes(submenuWrapperRef.current)) {
+      onCloseSubmenu();
+      document.removeEventListener('click', detectClickOutside);
+    }
+  }, [onCloseSubmenu]);
+
+  useEffect(() => {
+    if (isOpen) {
+      document.addEventListener('click', detectClickOutside);
+    }
+    if (submenu && isOpen) {
+      if (!didOpenByArrowUp && !didOpenByMouseEnter && !didOpenByClick) {
+        submenuItemsRefs.current[0].current.focus();
+      } else if (didOpenByArrowUp && !didOpenByMouseEnter && !didOpenByClick) {
+        submenuItemsRefs.current[submenuItemsRefs.current.length - 1].current.focus();
+      }
+    }
+  // eslint-disable-next-line max-len
+  }, [didOpenByArrowUp, submenu, submenuItemsRefs, detectClickOutside, isOpen, didOpenByMouseEnter, onCloseSubmenu, setIsOpenSubmenu, menuItemIndex, ref, didOpenByClick]);
+
+  const onKeyDownSubmenu = useCallback((ev, index) => {
+    if (!isOpen) {
+      if (Events.isEnterKey(ev) || Events.isSpaceKey(ev) || Events.isDownKey(ev)) {
+        ev.preventDefault();
+        setIsOpenSubmenu(menuItemIndex);
+      }
+      if (Events.isUpKey(ev)) {
+        ev.preventDefault();
+        setOpenByArrowUp(true);
+        setIsOpenSubmenu(menuItemIndex);
+      }
+    }
+
+    if (isOpen) {
+      if (didOpenByClick) {
+        if (Events.isDownKey(ev)) {
+          ev.preventDefault();
+          setOpenByClick(false);
+          setFocusToSubmenuElement(ev, 0);
+        }
+
+        if (Events.isUpKey(ev)) {
+          setOpenByArrowUp(true);
+          ev.preventDefault();
+          setOpenByClick(false);
+          setFocusToSubmenuElement(ev, submenuItemsRefs.current.length - 1);
+        }
+      }
+      if (Events.isDownKey(ev) && !didOpenByClick) {
+        ev.preventDefault();
+        setFocusToSubmenuElement(ev, index + 1);
+      }
+      if (Events.isUpKey(ev) && !didOpenByClick) {
+        ev.preventDefault();
+        setFocusToSubmenuElement(ev, index - 1);
+      }
+      if (
+        Events.isLeftKey(ev)
+        || Events.isRightKey(ev)
+        || Events.isTabKey(ev)
+      ) {
+        onCloseSubmenu();
+      }
+      if (Events.isHomeKey(ev)) {
+        ev.preventDefault();
+        submenuItemsRefs.current[0].current.focus();
+      }
+      if (Events.isEndKey(ev)) {
+        ev.preventDefault();
+        submenuItemsRefs.current[submenuItemsRefs.current.length - 1].current.focus();
+      }
+      if (Events.isEscKey(ev)) {
+        onCloseSubmenu();
+        ref.current.focus();
+      }
+      if (Events.isAlphabetKey(ev)) {
+        // A-Za-z: focus the next item on the list that starts with the pressed key
+        // selection should wrap to the start of the list
+        ev.stopPropagation();
+        let firstMatch;
+        let nextMatch;
+
+        React.Children.forEach(children, ({ props }, i) => {
+          if (props.children && props.children.toString().toLowerCase().startsWith(ev.key.toLowerCase())) {
```

Comment:
There is a small bug where the order is not correct on first load.
In this example pressing `i` once should move the focus down to the second item. Currently the first two presses of `i` has no effect.
![2020-08-12 10 15 40](https://user-images.githubusercontent.com/2328042/90039393-77cc5c00-dcbe-11ea-971e-996069e2f5c4.gif)